src/box/alter.cc

```diff
@@ -3927,11 +3927,36 @@ on_replace_dd_schema(struct trigger * /* trigger */, void *event)
 					      BOX_SCHEMA_FIELD_KEY);
 	if (key == NULL)
 		return -1;
-	if (strcmp(key, "cluster") == 0) {
+	if (strcmp(key, "cluster") == 0 ||
+	    strcmp(key, "replicaset_uuid") == 0) {
+		/*
+		 * Ignore old 'cluster' if it was already replaced by the new
+		 * key 'replicaset_uuid'.
+		 */
+		char mpkey[64];
+		char *mpkey_end = mp_encode_array(mpkey, 1);
+		mpkey_end = mp_encode_str0(mpkey_end, "replicaset_uuid");
+		struct tuple *other = NULL;
+		if (box_index_get(BOX_SCHEMA_ID, 0, mpkey, mpkey_end,
+				  &other) != 0)
+			return -1;
+		bool is_new_version_stored = other != NULL;
 		if (new_tuple == NULL) {
-			diag_set(ClientError, ER_REPLICASET_UUID_IS_RO);
-			return -1;
+			/*
+			 * Don't allow to delete the old key until the new one
+			 * is here. Otherwise recovery wouldn't work in case
+			 * the instance goes down after old deletion but before
+			 * new insertion.
+			 */
+			if (strcmp(key, "replicaset_uuid") == 0 ||
+			    !is_new_version_stored) {
+				diag_set(ClientError, ER_REPLICASET_UUID_IS_RO);
+				return -1;
+			}
+			return 0;
 		}
+		if (strcmp(key, "cluster") == 0 && is_new_version_stored)
+			return 0;
```

Comment:
It's rather difficult to understand what we're trying to achieve here. This deserves a comment. Anyway, I don't understand why we can't just allow both `cluster` and `replicaset_uuid` keys and handle them exactly in the same way, without any extra checks. Or, even better, check `dd_version_id` and use "replicaset_uuid" key name if it's newer than 3.0 or "cluster" otherwise.

Reply:
> handle them exactly in the same way

Done.

> why we can't just allow both

Now we do. Initially I wanted to disable that because `cluster` key is deprecated. They are not supposed to stay both.

> check dd_version_id

That won't work. Otherwise downgrade is not possible. In the new version I handle both keys and only check that at least one of them should always be present.